dbms/src/Storages/DeltaMerge/File/DMFile.cpp

```diff
@@ -701,5 +760,205 @@ void DMFile::initializeIndices()
+
+        auto checksum_result = digest->raw();
+        writeString(checksum_result.data(), checksum_result.size(), buffer);
+    }
+
+    MetaFooter footer{};
+    if (configuration)
+    {
+        footer.checksum_algorithm = static_cast<UInt64>(configuration->getChecksumAlgorithm());
+        footer.checksum_frame_length = configuration->getChecksumFrameLength();
+    }
+    writePODBinary(footer, buffer);
+}
+
+std::vector<char> DMFile::readMetaV2(const FileProviderPtr & file_provider)
+{
+    auto rbuf = openForRead(file_provider, metav2Path(), encryptionMetav2Path(), meta_buffer_size);
+    std::vector<char> buf(meta_buffer_size);
+    size_t read_bytes = 0;
+    for (;;)
+    {
+        read_bytes += rbuf.readBig(buf.data() + read_bytes, meta_buffer_size);
+        if (likely(read_bytes < buf.size()))
+        {
+            break;
+        }
+        LOG_WARNING(log, "{}'s size is larger than {}", metav2Path(), buf.size());
+        buf.resize(buf.size() + meta_buffer_size);
+    }
+    buf.resize(read_bytes);
+    return buf;
+}
+
+void DMFile::parseMetaV2(std::string_view buffer)
+{
+    // MetaFooter
+    const auto * footer = reinterpret_cast<const MetaFooter *>(buffer.data() + buffer.size() - sizeof(MetaFooter));
+    if (footer->checksum_algorithm != 0 && footer->checksum_frame_length != 0)
+    {
+        configuration = DMChecksumConfig{/*embedded_checksum*/ {}, footer->checksum_frame_length, static_cast<ChecksumAlgo>(footer->checksum_algorithm)};
+    }
+    else
+    {
+        configuration.reset();
+    }
+
+    const auto * ptr = reinterpret_cast<const char *>(footer);
+
+    // Checksum
+    if (configuration)
+    {
+        auto digest = configuration->createUnifiedDigest();
+        auto hash_size = digest->hashSize();
+        ptr = ptr - hash_size;
+        digest->update(buffer.data(), buffer.size() - sizeof(MetaFooter) - hash_size);
+        if (unlikely(digest->compareRaw(ptr)))
+        {
+            LOG_ERROR(log, "{} checksum invalid", metav2Path());
+            throw Exception(ErrorCodes::CORRUPTED_DATA, "{} checksum invalid", metav2Path());
+        }
+    }
+
+    ptr = ptr - sizeof(DMFileFormat::Version);
+    version = *(reinterpret_cast<const DMFileFormat::Version *>(ptr));
+
+    {
+        ptr = ptr - sizeof(MetaBlockHandle);
+        const auto * handle = reinterpret_cast<const MetaBlockHandle *>(ptr);
+        parseColumnStat(buffer.substr(handle->offset, handle->size));
+    }
+
+    {
+        ptr = ptr - sizeof(MetaBlockHandle);
+        const auto * handle = reinterpret_cast<const MetaBlockHandle *>(ptr);
+        parsePackProperty(buffer.substr(handle->offset, handle->size));
+    }
+    {
+        ptr = ptr - sizeof(MetaBlockHandle);
+        const auto * handle = reinterpret_cast<const MetaBlockHandle *>(ptr);
+        parsePackStat(buffer.substr(handle->offset, handle->size));
+    }
```

Comment:
Maybe add a `UInt64 block_type` before each MetaBlockHandle to indicate which type this block is?
Adding this we can add more type of optional "Block" in the future without changing the file format.